### PR TITLE
concurrent composition

### LINKED ISSUE #681: concurrent composition

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #681

This also adjusts the handling of errors in the interactive compositor: errors will bubble before popping a privacy parameter off, as errors are data-independent by the definition of measurements.

### CODE REVIEW

**Comment by Shoeboxam** [comment]
Current dependencies on/for this PR:
* main
  * **PR #702** <a href="https://app.graphite.dev/github/pr/opendp/opendp/702" target="_blank">Graphite</a> 
    * **PR #753** <a href="https://app.graphite.dev/github/pr/opendp/opendp/753" target="_blank">Graphite</a> 
      * **PR #703** <a href="https://app.graphite.dev/github/pr/opendp/opendp/703" target="_blank">Graphite</a> 
        * **PR #704** <a href="https://app.graphite.dev/github/pr/opendp/opendp/704" target="_blank">Graphite</a> 
          * **PR #754** <a href="https://app.graphite.dev/github/pr/opendp/opendp/754" target="_blank">Graphite</a> 
            * **PR #705** <a href="https://app.graphite.dev/github/pr/opendp/opendp/705" target="_blank">Graphite</a> 
              * **PR #874** <a href="https://app.graphite.dev/github/pr/opendp/opendp/874" target="_blank">Graphite</a> 
                * **PR #958** <a href="https://app.graphite.dev/github/pr/opendp/opendp/958" target="_blank">Graphite</a>  👈
                * **PR #907** <a href="https://app.graphite.dev/github/pr/opendp/opendp/907" target="_blank">Graphite</a> 
                  * **PR #916** <a href="https://app.graphite.dev/github/pr/opendp/opendp/916" target="_blank">Graphite</a> 
                    * **PR #917** <a href="https://app.graphite.dev/github/pr/opendp/opendp/917" target="_blank">Graphite</a> 
                      * **PR #921** <a href="https://app.graphite.dev/github/pr/opendp/opendp/921" target="_blank">Graphite</a> 
                      * **PR #763** <a href="https://app.graphite.dev/github/pr/opendp/opendp/763" target="_blank">Graphite</a> 
                        * **PR #765** <a href="https://app.graphite.dev/github/pr/opendp/opendp/765" target="_blank">Graphite</a> 
                          * **PR #767** <a href="https://app.graphite.dev/github/pr/opendp/opendp/767" target="_blank">Graphite</a> 
                            * **PR #769** <a href="https://app.graphite.dev/github/pr/opendp/opendp/769" target="_blank">Graphite</a> 
                              * **PR #771** <a href="https://app.graphite.dev/github/pr/opendp/opendp/771" target="_blank">Graphite</a> 
                                * **PR #862** <a href="https://app.graphite.dev/github/pr/opendp/opendp/862" target="_blank">Graphite</a> 
                                  * **PR #863** <a href="https://app.graphite.dev/github/pr/opendp/opendp/863" target="_blank">Graphite</a> 
                                    * **PR #865** <a href="https://app.graphite.dev/github/pr/opendp/opendp/865" target="_blank">Graphite</a> 
                                      * **PR #866** <a href="https://app.graphite.dev/github/pr/opendp/opendp/866" target="_blank">Graphite</a> 
                                        * **PR #774** <a href="https://app.graphite.dev/github/pr/opendp/opendp/774" target="_blank">Graphite</a> 
                                          * **PR #847** <a href="https://app.graphite.dev/github/pr/opendp/opendp/847" target="_blank">Graphite</a> 
                                            * **PR #795** <a href="https://app.graphite.dev/github/pr/opendp/opendp/795" target="_blank">Graphite</a> 
                                              * **PR #797** <a href="https://app.graphite.dev/github/pr/opendp/opendp/797" target="_blank">Graphite</a> 
                                                * **PR #824** <a href="https://app.graphite.dev/github/pr/opendp/opendp/824" target="_blank">Graphite</a> 
                                                  * **PR #841** <a href="https://app.graphite.dev/github/pr/opendp/opendp/841" target="_blank">Graphite</a> 
                                                    * **PR #896** <a href="https://app.graphite.dev/github/pr/opendp/opendp/896" target="_blank">Graphite</a> 
                                                    * **PR #868** <a href="https://app.graphite.dev/github/pr/opendp/opendp/868" target="_blank">Graphite</a> 
                                                      * **PR #864** <a href="https://app.graphite.dev/github/pr/opendp/opendp/864" target="_blank">Graphite</a> 
                                                      * **PR #819** <a href="https://app.graphite.dev/github/pr/opendp/opendp/819" target="_blank">Graphite</a> 
                                                        * **PR #829** <a href="https://app.graphite.dev/github/pr/opendp/opendp/829" target="_blank">Graphite</a> 
                                                          * **PR #890** <a href="https://app.graphite.dev/github/pr/opendp/opendp/890" target="_blank">Graphite</a> 
                                                            * **PR #856** <a href="https://app.graphite.dev/github/pr/opendp/opendp/856" target="_blank">Graphite</a> 
                                                              * **PR #908** <a href="https://app.graphite.dev/github/pr/opendp/opendp/908" target="_blank">Graphite</a> 
                                                                * **PR #913** <a href="https://app.graphite.dev/github/pr/opendp/opendp/913" target="_blank">Graphite</a> 
                                                                  * **PR #952** <a href="https://app.graphite.dev/github/pr/opendp/opendp/952" target="_blank">Graphite</a> 
                                                                    * **PR #954** <a href="https://app.graphite.dev/github/pr/opendp/opendp/954" target="_blank">Graphite</a> 
                                                                  * **PR #915** <a href="https://app.graphite.dev/github/pr/opendp/opendp/915" target="_blank">Graphite</a> 
                                                          * **PR #888** <a href="https://app.graphite.dev/github/pr/opendp/opendp/888" target="_blank">Graphite</a> 
                                                          * **PR #854** <a href="https://app.graphite.dev/github/pr/opendp/opendp/854" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/opendp/opendp/958?utm_source=stack-comment).

File: rust/src/combinators/sequential_compositor/noninteractive/mod.rs [hunk 1]
```diff
@@ -97,16 +102,23 @@ where
 }
 
 pub trait BasicCompositionMeasure: Measure {
+    fn concurrent(&self) -> bool;
```
Comment (andrewvyrros) [inline rust/src/combinators/sequential_compositor/noninteractive/mod.rs:105]: Will probably need this to be Fallible at some point.

File: rust/src/combinators/sequential_compositor/noninteractive/mod.rs [hunk 2]
```diff
@@ -27,7 +30,7 @@ use crate::{
 /// * `MI` - Input Metric
 /// * `MO` - Output Metric
```
Comment (andrewvyrros) [inline rust/src/combinators/sequential_compositor/noninteractive/mod.rs:31]: Output Measure

File: rust/src/combinators/sequential_compositor/noninteractive/mod.rs [hunk 3]
```diff
@@ -66,23 +69,25 @@ where
...
-                        .map(|f| f.eval(arg))
-                        .collect::<Fallible<_>>()
-                },
-            )
+            let go = || functions.iter().map(|f| f.eval(arg)).collect();
+
+            if concurrent {
+                go()
+            } else {
+                wrap(
```
Comment (andrewvyrros) [inline rust/src/combinators/sequential_compositor/noninteractive/mod.rs:81]: Very cool to use wrapping here to enforce concurrency where needed.

File: rust/src/combinators/sequential_compositor/interactive/ffi.rs [hunk 4]
```diff
@@ -19,7 +19,19 @@ use crate::{
...
+/// **Composition Properties**
+///
+/// * sequential: all measurements are applied to the same dataset
+/// * basic: the composition is the linear sum of the privacy usage of each query
+/// * interactive: mechanisms can be specified based on answers to previous queries
+/// * compositor: all privacy parameters specified up-front
+///
+/// If the privacy measure supports concurrency,
+/// this compositor allows you to spawn multiple interactive mechanisms
+/// and interleave your queries amongst them.
```
Comment (mccalluc) [inline rust/src/combinators/sequential_compositor/interactive/ffi.rs:34]: I worry a bit whenever things get duplicated. Consider trimming one of these down, referencing the other, and explaining how they are different? (I'm still wondering about how we publish docs for private elements... It does add a lot more for the end user to swim through, and makes it harder to use search to find the right method.)

**Review by mccalluc — COMMENTED** [review]
Feeling out of my depth here, but let me know if you'd like to walk me through part of it.

File: rust/src/combinators/sequential_compositor/interactive/ffi.rs (hunk 4 above)
Comment (Shoeboxam) [reply rust/src/combinators/sequential_compositor/interactive/ffi.rs:34]: I wrote this PR before you could add annotations to bootstrap to suppress generic types in the codegen.  I've adjusted the PR to apply bootstrap to the original function, which avoids needing multiple copies of the docs.

**Review by mccalluc — APPROVED** [review]
One documentation question; Will file follow-up issue. Approved.

File: python/src/opendp/combinators.py [hunk 5]
```diff
@@ -31,7 +32,14 @@ def make_basic_composition(
...
     
-    All metrics and domains must be equivalent, except for the output domain.
+    All metrics and domains must be equivalent.
+    
+    **Composition Properties**
+    
+    * sequential: all measurements are applied to the same dataset
+    * basic: the composition is the linear sum of the privacy usage of each query
+    * noninteractive: all mechanisms specified up-front (but each can be interactive)
+    * compositor: all privacy parameters specified up-front (via the map)
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply python/src/opendp/combinators.py:42]: I don't quite understand the question, this is not a supporting elements section.

**Pull request merged**